### PR TITLE
Create opaque definitions in resolver.

### PR DESCRIPTION
**Pull request by cjgillot**
Implementing https://github.com/rust-lang/rust/issues/129023#issuecomment-2306079532

That was easier than I expected.

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_ast/src/visit.rs [hunk 1]
```diff
@@ -69,14 +69,14 @@ pub enum FnKind<'a> {
...
-    Closure(&'a ClosureBinder, &'a FnDecl, &'a Expr),
+    Closure(&'a ClosureBinder, &'a Option<CoroutineKind>, &'a FnDecl, &'a Expr),
 }
 
 impl<'a> FnKind<'a> {
     pub fn header(&self) -> Option<&'a FnHeader> {
         match *self {
             FnKind::Fn(_, _, sig, _, _, _) => Some(&sig.header),
-            FnKind::Closure(_, _, _) => None,
+            FnKind::Closure(_, _, _, _) => None,
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/visit.rs:79]:
```suggestion
            FnKind::Closure(..) => None,
```

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 2]
```diff
@@ -1400,27 +1399,12 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
-                            DefKind::TyParam,
-                            span,
-                        );
+                        let def_id = self.local_def_id(*def_node_id);
+                        let name = self.tcx.item_name(def_id.to_def_id());
+                        let ident = Ident::new(name, span);
                         let (param, bounds, path) = self.lower_universal_param_and_bounds(
                             *def_node_id,
-                            span,
+                            t.span,
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/lib.rs:1407]:
```suggestion
                            span,
```
Nit: it's the same span as before.

File: compiler/rustc_resolve/src/def_collector.rs [hunk 3]
```diff
@@ -183,38 +182,55 @@ impl<'a, 'b, 'tcx> visit::Visitor<'a> for DefCollector<'a, 'b, 'tcx> {
...
-                        let closure_def = self.create_def(
-                            coroutine_kind.closure_id(),
-                            kw::Empty,
-                            DefKind::Closure,
-                            span,
-                        );
-                        self.with_parent(closure_def, |this| this.visit_block(body));
-                    }
-                    return;
+        let walk_fn_decl = |this: &mut Self,
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/def_collector.rs:185]: Why is this a closure?
It's only called once.

**Comment by petrochenkov** [comment]
r=me after squashing.
@rustbot author

File: compiler/rustc_resolve/src/def_collector.rs [hunk 4]
```diff
@@ -334,27 +346,7 @@ impl<'a, 'b, 'tcx> visit::Visitor<'a> for DefCollector<'a, 'b, 'tcx> {
...
-                            );
-                            this.with_parent(coroutine_def, |this| visit::walk_expr(this, expr));
-                        });
-                        return;
-                    }
-                    None => closure_def,
-                }
-            }
-            ExprKind::Gen(_, _, _, _) => {
+            ExprKind::Closure(..) | ExprKind::Gen(..) => {
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/def_collector.rs:349]: FYI, I simplified the `Closure` arm to just create the def-id, like other branches do. There is no additional behaviour change.

**Pull request merged**